File: src/CommandLineCommands/ChatCommand.cs
using Microsoft.Extensions.AI;
using System.Text;
using System.Threading.Tasks;
using System.Collections.Generic;
using System.Linq;
using chatx.FunctionCalling;
using chatx.McpHelpers;
using ModelContextProtocol.Client;

public class ChatCommand : Command
{
    public ChatCommand()
    {
    }

    override public bool IsEmpty()
    {
        return false;
    }

    override public string GetCommandName()
    {
        return "";
    }

    public ChatCommand Clone()
    {
        var clone = new ChatCommand();
        
        // Copy all properties
        clone.SystemPrompt = this.SystemPrompt;
        clone.SystemPromptAdds = new List<string>(this.SystemPromptAdds);
        clone.UserPromptAdds = new List<string>(this.UserPromptAdds);
        clone.TrimTokenTarget = this.TrimTokenTarget;
        clone.MaxOutputTokens = this.MaxOutputTokens;
        clone.LoadMostRecentChatHistory = this.LoadMostRecentChatHistory;
        clone.InputChatHistory = this.InputChatHistory;
        clone.OutputChatHistory = this.OutputChatHistory;
        clone.OutputTrajectory = this.OutputTrajectory;
        clone.InputInstructions = new List<string>(this.InputInstructions);
        clone.UseTemplates = this.UseTemplates;
        
        // Deep copy variables dictionary
        clone.Variables = new Dictionary<string, string>(this.Variables);
        
        return clone;
    }

    public async Task<int> ExecuteAsync(bool interactive)
    {
        // Setup the named values
        _namedValues = new TemplateVariables(Variables);

        // Transfer known settings to the command if not already set
        var maxOutputTokensSetting = ConfigStore.Instance.GetFromAnyScope(KnownSettings.AppMaxTokens);
        var useMaxOutputTokenSetting = !MaxOutputTokens.HasValue && maxOutputTokensSetting.AsInt() > 0;
        if (useMaxOutputTokenSetting) MaxOutputTokens = maxOutputTokensSetting.AsInt();

        // Ground the filenames (in case they're templatized, or auto-save is enabled).
        InputChatHistory = ChatHistoryFileHelpers.GroundInputChatHistoryFileName(InputChatHistory, LoadMostRecentChatHistory)?.ReplaceValues(_namedValues);
        OutputChatHistory = ChatHistoryFileHelpers.GroundOutputChatHistoryFileName(OutputChatHistory)?.ReplaceValues(_namedValues);
        OutputTrajectory = ChatHistoryFileHelpers.GroundOutputTrajectoryFileName(OutputTrajectory)?.ReplaceValues(_namedValues);
        _trajectoryFile = new TrajectoryFile(OutputTrajectory);

        // Ground the system prompt, added user messages, and InputInstructions.
        SystemPrompt = GroundSystemPrompt();
        UserPromptAdds = GroundUserPromptAdds();
        InputInstructions = GroundInputInstructions();

        // Create the function factory and add functions.
        var factory = new McpFunctionFactory();
        factory.AddFunctions(new DateAndTimeHelperFunctions());
        factory.AddFunctions(new ShellCommandToolHelperFunctions());
        factory.AddFunctions(new StrReplaceEditorHelperFunctions());
        factory.AddFunctions(new ThinkingToolHelperFunction());
        
        // Add MCP functions if any are configured
        await AddMcpFunctions(factory);
        factory.AddFunctions(new CodeExplorationHelperFunctions());

        // Create the chat completions object with the external ChatClient and system prompt.
        var chatClient = ChatClientFactory.CreateChatClient();
        var chat = new FunctionCallingChat(chatClient, SystemPrompt, factory, MaxOutputTokens);

        try
        {
            // Add the user prompt messages to the chat.
            chat.AddUserMessages(UserPromptAdds);

            // Load the chat history from the file.
            var loadChatHistory = !string.IsNullOrEmpty(InputChatHistory);
            if (loadChatHistory) chat.LoadChatHistory(InputChatHistory!, TrimTokenTarget ?? DefaultTrimTokenTarget, useOpenAIFormat: ChatHistoryDefaults.UseOpenAIFormat);

            // Check to make sure we're either in interactive mode, or have input instructions.
            if (!interactive && InputInstructions.Count == 0)
            {
                ConsoleHelpers.WriteWarning("\nNo input instructions provided. Exiting.");
                return 1;
            }

            while (true)
            {
                DisplayUserPrompt();
                var userPrompt = interactive
                    ? InteractivelyReadLineOrSimulateInput(InputInstructions, "exit")
                    : ReadLineOrSimulateInput(InputInstructions, "exit");
                if (string.IsNullOrWhiteSpace(userPrompt) || userPrompt == "exit") break;

                var (skipAssistant, replaceUserPrompt) = await TryHandleChatCommandAsync(chat, userPrompt);
                if (skipAssistant) continue; // Some chat commands don't require a response from the assistant.

                var shouldReplaceUserPrompt = !string.IsNullOrEmpty(replaceUserPrompt);
                if (shouldReplaceUserPrompt) DisplayPromptReplacement(userPrompt, replaceUserPrompt);

                var giveAssistant = shouldReplaceUserPrompt ? replaceUserPrompt! : userPrompt;

                DisplayAssistantLabel();
                var response = await CompleteChatStreamingAsync(chat, giveAssistant,
                    (messages) => HandleUpdateMessages(messages),
                    (update) => HandleStreamingChatCompletionUpdate(update),
                    (name, args, result) => HandleFunctionCallCompleted(name, args, result));
                ConsoleHelpers.WriteLine("\n", overrideQuiet: true);
            }

            return 0;
        }
        finally
        {
            // Clean up resources
            await chat.DisposeAsync();
            
            // If we have an MCP function factory, dispose its resources too
            if (factory is McpFunctionFactory mcpFactory)
            {
                await mcpFactory.DisposeAsync();
            }
        }
    }

    private string GroundSystemPrompt()
    {
        SystemPrompt ??= GetBuiltInSystemPrompt();

        var processed =  ProcessTemplate(SystemPrompt + "\n\n" + GetSystemPromptAdds());
        return _namedValues != null ? processed.ReplaceValues(_namedValues) : processed;
    }

    private List<string> GroundUserPromptAdds()
    {
        return UserPromptAdds
            .Select(x => UseTemplates ? ProcessTemplate(x) : x)
            .Select(x => _namedValues != null ? x.ReplaceValues(_namedValues) : x)
            .ToList();
    }

    private List<string> GroundInputInstructions()
    {
        return InputInstructions
            .Select(x => UseTemplates ? ProcessTemplate(x) : x)
            .Select(x => _namedValues != null ? x.ReplaceValues(_namedValues) : x)
            .ToList();
    }

    private string GetBuiltInSystemPrompt()
    {
        if (EmbeddedFileHelpers.EmbeddedStreamExists("prompts.system.md"))
        {
            var text = EmbeddedFileHelpers.ReadEmbeddedStream("prompts.system.md")!;
            return ProcessTemplate(text);
        }

        return "You are a helpful AI assistant.";
    }

    private string GetSystemPromptAdds()
    {
        var processedAdds = SystemPromptAdds
            .Select(x => UseTemplates ? ProcessTemplate(x) : x)
            .ToList();
        var joined = string.Join("\n\n", processedAdds);
        return joined.Trim(new char[] { '\n', '\r', ' ' });
    }

    private async Task<(bool skipAssistant, string? giveAssistant)> TryHandleChatCommandAsync(FunctionCallingChat chat, string userPrompt)
    {
        bool skipAssistant = false;
        string? giveAssistant = null;

        if (userPrompt.StartsWith("/save"))
        {
            skipAssistant = HandleSaveChatHistoryCommand(chat, userPrompt.Substring("/save".Length).Trim());
        }
        else if (userPrompt == "/clear")
        {
            skipAssistant = HandleClearChatHistoryCommand(chat);
        }
        else if (userPrompt == "/cost")
        {
            skipAssistant = HandleShowCostCommand();
        }
        else if (userPrompt == "/help")
        {
            skipAssistant = HandleHelpCommand();
        }
        else if (_mdxCommandHandler.IsCommand(userPrompt))
        {
            skipAssistant = await HandleMdxCommand(chat, userPrompt);
        }
        else if (_promptCommandHandler.IsCommand(userPrompt))
        {
            var handled = HandlePromptCommand(chat, userPrompt, out giveAssistant);
            if (!handled) giveAssistant = null;
        }

        return (skipAssistant, giveAssistant);
    }

    private void DisplayPromptReplacement(string userPrompt, string? replaceUserPrompt)
    {
        ConsoleHelpers.WriteLine($"\rUser: {userPrompt} => {replaceUserPrompt}", ConsoleColor.DarkGray, overrideQuiet: true);
    }

    private bool HandlePromptCommand(FunctionCallingChat chat, string userPrompt, out string? giveAssistant)
    {
        giveAssistant = _promptCommandHandler.HandleCommand(userPrompt);
        return !string.IsNullOrEmpty(giveAssistant);
    }

    private async Task<bool> HandleMdxCommand(FunctionCallingChat chat, string userPrompt)
    {
        var userFunctionName = _mdxCommandHandler.GetCommandName(userPrompt);
        DisplayUserFunctionCall(userFunctionName, null);

        var result = await _mdxCommandHandler.HandleCommand(userPrompt);
        if (result != null) chat.AddUserMessage(result);

        DisplayUserFunctionCall(userFunctionName, result ?? string.Empty);
        return true;
    }

    private bool HandleClearChatHistoryCommand(FunctionCallingChat chat)
    {
        chat.ClearChatHistory();
        _totalTokensIn = 0;
        _totalTokensOut = 0;
        ConsoleHelpers.WriteLine("Cleared chat history.\n", ConsoleColor.Yellow, overrideQuiet: true);
        return true;
    }

    private bool HandleSaveChatHistoryCommand(FunctionCallingChat chat, string? fileName = null)
    {
        var useDefaultFileName = string.IsNullOrEmpty(fileName);
        if (useDefaultFileName) fileName = "chat-history.jsonl";

        ConsoleHelpers.Write($"Saving {fileName} ...", ConsoleColor.Yellow, overrideQuiet: true);
        chat.SaveChatHistoryToFile(fileName!, useOpenAIFormat: ChatHistoryDefaults.UseOpenAIFormat);
        ConsoleHelpers.WriteLine("Saved!\n", ConsoleColor.Yellow, overrideQuiet: true);

        return true;
    }

    private bool HandleShowCostCommand()
    {
        ConsoleHelpers.WriteLine($"Tokens: {_totalTokensIn} in, {_totalTokensOut} out\n", ConsoleColor.Yellow, overrideQuiet: true);
        return true;
    }

    private bool HandleHelpCommand()
    {
        var helpBuilder = new StringBuilder();
        
        // Built-in chat commands
        helpBuilder.AppendLine("BUILT-IN");
        helpBuilder.AppendLine();
        helpBuilder.AppendLine("  /save     Save chat history to file");
        helpBuilder.AppendLine("  /clear    Clear chat history");
        helpBuilder.AppendLine("  /cost     Show token usage statistics");
        helpBuilder.AppendLine("  /help     Show this help message");
        helpBuilder.AppendLine();
        
        // MDX integration commands
        helpBuilder.AppendLine("EXTERNAL");
        helpBuilder.AppendLine();
        helpBuilder.AppendLine("  /files    List files matching pattern");
        helpBuilder.AppendLine("  /file     Get contents of a file");
        helpBuilder.AppendLine("  /find     Find content in files");
        helpBuilder.AppendLine();
        helpBuilder.AppendLine("  /search   Search the web");
        helpBuilder.AppendLine("  /get      Get content from URL");
        helpBuilder.AppendLine();
        helpBuilder.AppendLine("  /run      Run a command");
        helpBuilder.AppendLine();

        // User-defined prompts
        helpBuilder.AppendLine("PROMPTS");
        helpBuilder.AppendLine();
        
        // Get all scopes for prompt files
        bool foundPrompts = false;
        foreach (var scope in new[] { ConfigFileScope.Local, ConfigFileScope.User, ConfigFileScope.Global })
        {
            var promptDir = PromptFileHelpers.FindPromptDirectoryInScope(scope);
            if (promptDir == null || !Directory.Exists(promptDir)) continue;
            
            var promptFiles = Directory.GetFiles(promptDir, "*.prompt")
                .OrderBy(file => Path.GetFileNameWithoutExtension(file))
                .ToList();
                
            if (promptFiles.Count == 0) continue;
            
            foundPrompts = true;
            
            // Format location header consistently
            var locationDisplay = CommonDisplayHelpers.FormatLocationHeader(promptDir, scope);
            helpBuilder.AppendLine($"  LOCATION: {locationDisplay}");
            helpBuilder.AppendLine();

            var indent = new string(' ', 4);
            
            foreach (var promptFile in promptFiles)
            {
                var promptName = Path.GetFileNameWithoutExtension(promptFile);
                helpBuilder.AppendLine($"{indent}/{promptName}");
                helpBuilder.AppendLine();

                // Get prompt content and truncate it for preview
                var promptText = PromptFileHelpers.GetPromptText(promptName, scope)!;
                var truncatedContent = CommonDisplayHelpers.TruncateContent(
                    promptText, 
                    CommonDisplayHelpers.DefaultMaxContentLines, 
                    CommonDisplayHelpers.DefaultMaxContentWidth, 
                    8);  // deeper indentation for content
                
                helpBuilder.AppendLine(truncatedContent);
                helpBuilder.AppendLine();
            }
            
            helpBuilder.AppendLine();
        }
        
        if (!foundPrompts)
        {
            helpBuilder.AppendLine("  No custom prompts found.");
            helpBuilder.AppendLine();
        }
        
        var indented = "\n  " + helpBuilder.ToString().Replace("\n", "\n  ").TrimEnd() + "\n";
        ConsoleHelpers.WriteLine(indented, overrideQuiet: true);
        return true;
    }

    private async Task<string> CompleteChatStreamingAsync(
        FunctionCallingChat chat,
        string userPrompt,
        Action<IList<ChatMessage>>? messageCallback = null,
        Action<ChatResponseUpdate>? streamingCallback = null,
        Action<string, string, string?>? functionCallCallback = null)
    {
        messageCallback = TryCatchHelpers.NoThrowWrap(messageCallback);
        streamingCallback = TryCatchHelpers.NoThrowWrap(streamingCallback);
        functionCallCallback = TryCatchHelpers.NoThrowWrap(functionCallCallback);

        try
        {
            var response = await chat.CompleteChatStreamingAsync(userPrompt,
                (messages) => messageCallback?.Invoke(messages),
                (update) => streamingCallback?.Invoke(update),
                (name, args, result) => functionCallCallback?.Invoke(name, args, result));

            return response;
        }
        catch (Exception)
        {
            SaveExceptionHistory(chat);
            throw;
        }
    }

    private string? ReadLineOrSimulateInput(List<string> inputInstructions, string? defaultOnEndOfInput = null)
    {
        while (inputInstructions?.Count > 0)
        {
            var input = inputInstructions[0];
            inputInstructions.RemoveAt(0);

            var empty = string.IsNullOrEmpty(input);
            if (!empty)
            {
                ConsoleHelpers.WriteLine(input);
                return input;
            }
        }

        if (Console.IsInputRedirected)
        {
            var line = Console.ReadLine();
            line ??= defaultOnEndOfInput;
            if (line != null) ConsoleHelpers.WriteLine(line);
            return line;
        }

        return defaultOnEndOfInput;
    }

    private string? InteractivelyReadLineOrSimulateInput(List<string> inputInstructions, string? defaultOnEndOfInput = null)
    {
        var input = ReadLineOrSimulateInput(inputInstructions, null);
        if (input != null) return input;

        return Console.ReadLine() ?? defaultOnEndOfInput;
    }

    private void HandleUpdateMessages(IList<ChatMessage> messages)
    {
        messages.TryTrimToTarget(TrimTokenTarget ?? DefaultTrimTokenTarget);

        if (OutputChatHistory != null)
        {
            messages.SaveChatHistoryToFile(OutputChatHistory, useOpenAIFormat: ChatHistoryDefaults.UseOpenAIFormat);
        }
        
        _trajectoryFile?.AppendMessage(messages.LastOrDefault());
    }

    private void HandleStreamingChatCompletionUpdate(ChatResponseUpdate update)
    {
        var usageUpdate = update.Contents
            .Where(x => x is UsageContent)
            .Cast<UsageContent>()
            .FirstOrDefault();
        var inTokens = usageUpdate?.Details.InputTokenCount ?? 0;
        var outTokens = usageUpdate?.Details.OutputTokenCount ?? 0;
        if (inTokens > 0 || outTokens > 0)
        {
            _totalTokensIn += inTokens;
            _totalTokensOut += outTokens;
            if (ConsoleHelpers.IsVerbose())
            {
                ConsoleHelpers.WriteLine($"\nTokens: {inTokens} in ({_totalTokensIn} total), {outTokens} out ({_totalTokensOut} total)", ConsoleColor.DarkYellow);
            }
        }

        var text = string.Join("", update.Contents
            .Where(x => x is TextContent)
            .Cast<TextContent>()
            .Select(x => x.Text)
            .ToList());
        DisplayAssistantResponse(text);
    }

    private void HandleFunctionCallCompleted(string name, string args, string? result)
    {
        DisplayAssistantFunctionCall(name, args, result);
    }

    private void DisplayUserPrompt()
    {
        ConsoleHelpers.Write("\rUser: ", ConsoleColor.Green);
        Console.ForegroundColor = ConsoleColor.White;
    }

    private void DisplayUserFunctionCall(string userFunctionName, string? result)
    {
        ConsoleHelpers.Write($"\ruser-function: {userFunctionName} => ", ConsoleColor.DarkGray);

        if (result == null) ConsoleHelpers.Write("...", ConsoleColor.DarkGray);
        if (result != null)
        {
            ConsoleHelpers.WriteLine(result, ConsoleColor.DarkGray);
            DisplayUserPrompt();
        }
    }

    private void DisplayAssistantLabel()
    {
        ConsoleHelpers.Write("\nAssistant: ", ConsoleColor.Green);
        _assistantResponseCharsSinceLabel = 0;
    }

    private void DisplayAssistantResponse(string text)
    {
        if (_assistantResponseCharsSinceLabel == 0 && text.StartsWith("\n"))
        {
            text = text.TrimStart(new char[] { '\n', '\r', ' ' });
        }

        ConsoleHelpers.Write(text, ConsoleColor.White, overrideQuiet: true);

        _assistantResponseCharsSinceLabel += text.Length;
        _asssistantResponseNeedsLF = !text.EndsWith("\n");
    }

    private void EnsureLineFeeds()
    {
        if (_asssistantResponseNeedsLF)
        {
            var oneLineFeedOrTwo = ConsoleHelpers.IsQuiet() ? "\n\n" : "\n";
            ConsoleHelpers.Write(oneLineFeedOrTwo, overrideQuiet: true);
            _asssistantResponseNeedsLF = false;
        }
    }

    private void DisplayAssistantFunctionCall(string name, string args, string? result)
    {
        EnsureLineFeeds();
        switch (name)
        {
            case "Think":
                DisplayAssistantThinkFunctionCall(args, result);
                break;

            default:
                DisplayGenericAssistantFunctionCall(name, args, result);
                break;
        }
    }
    
    private void DisplayAssistantThinkFunctionCall(string args, string? result)
    {
        var thought = JsonHelpers.GetJsonPropertyValue(args, "thought", args);
        var hasThought = !string.IsNullOrEmpty(thought);
        var hasResult = !string.IsNullOrEmpty(result);

        if (hasThought && !hasResult) ConsoleHelpers.WriteLine($"\n[THINKING]\n{thought}", ConsoleColor.DarkCyan);
        if (hasResult)
        {
            ConsoleHelpers.WriteLine($"\n{result}", ConsoleColor.DarkGray);
            DisplayAssistantLabel();
        }
    }
    
    private void DisplayGenericAssistantFunctionCall(string name, string args, string? result)
    {
        ConsoleHelpers.Write($"\rassistant-function: {name} {args} => ", ConsoleColor.DarkGray);
        
        if (result == null) ConsoleHelpers.Write("...", ConsoleColor.DarkGray);
        if (result != null)
        {
            ConsoleHelpers.WriteLine(result, ConsoleColor.DarkGray);
            DisplayAssistantLabel();
        }
    }

    private string ProcessTemplate(string template)
    {
        if (string.IsNullOrEmpty(template))
        {
            return template;
        }

        var variables = new TemplateVariables(Variables);
        return TemplateHelpers.ProcessTemplate(template, variables);
    }

    private static void SaveExceptionHistory(FunctionCallingChat chat)
    {
        ConsoleHelpers.Write("\n\n", overrideQuiet: true);
        SaveExceptionChatHistory(chat);
        ConsoleHelpers.Write("\n", overrideQuiet: true);
        SaveExceptionTrajectory(chat);
    }

    private static void SaveExceptionChatHistory(FunctionCallingChat chat)
    {
        try
        {
            var fileName = FileHelpers.GetFileNameFromTemplate("exception-chat-history.jsonl", "{filebase}-{time}.{fileext}")!;
            chat.SaveChatHistoryToFile(fileName, useOpenAIFormat: ChatHistoryDefaults.UseOpenAIFormat);
            ConsoleHelpers.WriteWarning($"SAVED: {fileName}");
        }
        catch (Exception)
        {
            ConsoleHelpers.WriteWarning("Failed to save exception chat history.");
        }
    }

    private static void SaveExceptionTrajectory(FunctionCallingChat chat)
    {
        try
        {
            var fileName = FileHelpers.GetFileNameFromTemplate("exception-trajectory.md", "{filebase}-{time}.{fileext}")!;
            chat.SaveTrajectoryToFile(fileName, useOpenAIFormat: ChatHistoryDefaults.UseOpenAIFormat);
            ConsoleHelpers.WriteWarning($"SAVED: {fileName}");
        }
        catch (Exception)
        {
            ConsoleHelpers.WriteWarning("Failed to save exception trajectory.");
        }
    }

    public string? SystemPrompt { get; set; }
    public List<string> SystemPromptAdds { get; set; } = new List<string>();
    public List<string> UserPromptAdds { get; set; } = new List<string>();

    public int? TrimTokenTarget { get; set; }
    public int? MaxOutputTokens { get; set; }

    public bool LoadMostRecentChatHistory = false;
    public string? InputChatHistory;
    public string? OutputChatHistory;
    public string? OutputTrajectory;

    public List<string> InputInstructions = new();
    public bool UseTemplates = true;

    public Dictionary<string, string> Variables { get; set; } = new Dictionary<string, string>();
    public List<ForEachVariable> ForEachVariables { get; set; } = new List<ForEachVariable>();

    private int _assistantResponseCharsSinceLabel = 0;
    private bool _asssistantResponseNeedsLF = false;

    private INamedValues? _namedValues;
    private TrajectoryFile? _trajectoryFile;
    private SlashMdxCommandHandler _mdxCommandHandler = new();
    private SlashPromptCommandHandler _promptCommandHandler = new();

    private long _totalTokensIn = 0;
    private long _totalTokensOut = 0;
    private const int DefaultTrimTokenTarget = 160000;

    private async Task AddMcpFunctions(McpFunctionFactory factory)
    {
        try
        {
            // Create clients for all configured MCP servers
            var clients = await McpClientManager.CreateAllClientsAsync();
            if (clients.Count == 0)
            {
                return; // No configured MCP servers
            }

            Console.WriteLine($"Found {clients.Count} MCP server(s)");
            
            // Add tools from each client
            foreach (var clientEntry in clients)
            {
                var serverName = clientEntry.Key;
                var client = clientEntry.Value;

                try
                {
                    await factory.AddMcpClientToolsAsync(client, serverName);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error adding tools from MCP server '{serverName}': {ex.Message}");
                }
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error loading MCP functions: {ex.Message}");
        }
    }
}

File: ts/src/commands/ChatCommand.ts
import { Command } from 'commander';
import { ChatMessage } from '../types';
import { ConsoleHelpers } from '../helpers/ConsoleHelpers';
import { FunctionCallingChat } from '../chat/FunctionCallingChat';
import { ChatClientFactory } from '../chat/ChatClientFactory';

export class ChatCommand {
  // Core properties matching C# implementation
  public systemPrompt?: string;
  public systemPromptAdds: string[] = [];
  public userPromptAdds: string[] = [];
  
  public maxPromptTokenTarget: number = 50000;  // DefaultMaxPromptTokenTarget
  public maxToolTokenTarget: number = 50000;    // DefaultMaxToolTokenTarget
  public maxOutputTokens?: number;
  public maxChatTokenTarget: number = 160000;   // DefaultMaxChatTokenTarget
  
  public loadMostRecentChatHistory: boolean = false;
  public inputChatHistory?: string;
  public outputChatHistory?: string;
  public outputTrajectory?: string;
  
  public autoSaveOutputChatHistory?: string;
  public autoSaveOutputTrajectory?: string;
  
  public inputInstructions: string[] = [];
  public useTemplates: boolean = true;
  
  public useMcps: string[] = [];
  public imagePatterns: string[] = [];
  public variables: Record<string, string> = {};
  
  // Private state matching C# implementation
  private assistantResponseCharsSinceLabel: number = 0;
  private assistantResponseNeedsLF: boolean = false;
  private totalTokensIn: number = 0;
  private totalTokensOut: number = 0;
  private approvedFunctionCallNames: Set<string> = new Set();
  private deniedFunctionCallNames: Set<string> = new Set();

  static createCommand(): Command {
    const chatCmd = new Command('chat');
    
    chatCmd
      .description('Start chat session with AI assistant')
      .option('--input <text>', 'Single input line or question')
      .option('--inputs <text...>', 'Chain multiple questions/requests')
      .option('-q, --question <text>', 'Shortcut for non-interactive, quiet mode')
      .option('--interactive [boolean]', 'Enable/disable interactive mode', true)
      .option('--quiet', 'Suppress verbose output')
      .option('--output-chat-history [filename]', 'Save conversation')
      .option('--input-chat-history <filename>', 'Load previous conversation')
      .option('--chat-history <filename>', 'Use same filename for input and output')
      .option('--continue', 'Resume most recent chat history')
      .option('--system-prompt <text>', 'Completely replace default system behavior')
      .option('--add-system-prompt <text>', 'Augment existing default system prompt')
      .option('--var <name=value>', 'Define a variable', (value, previous: Record<string, string>) => {
        const vars = previous || {};
        const [name, val] = value.split('=', 2);
        if (name && val !== undefined) {
          vars[name] = val;
        }
        return vars;
      }, {} as Record<string, string>)
      .option('--vars <name=value...>', 'Define multiple variables')
      .option('--foreach <spec>', 'Run command with different variable values in parallel')
      .option('--threads <number>', 'Specify maximum parallel processes', parseInt)
      .option('--profile <name>', 'Load specific configuration profile')
      .option('--max-output-tokens <number>', 'Maximum output tokens', parseInt)
      .action(async (options, command) => {
        const chatCommand = new ChatCommand();
        await chatCommand.execute(options, command);
      });

    return chatCmd;
  }

  async execute(commandOptions: any, command: Command): Promise<number> {
    try {
      // Process options and set properties
      this.processOptions(commandOptions);
      
      // Determine if we're in interactive mode
      const hasInputInstructions = this.inputInstructions.length > 0;
      const interactive = commandOptions.interactive && !hasInputInstructions;
      
      // Check if we have input when not in interactive mode
      if (!interactive && !hasInputInstructions) {
        ConsoleHelpers.writeWarning('\nNo input instructions provided. Exiting.');
        return 1;
      }

      // Create the chat client and function calling chat (matching C# ChatCommand.cs:99-100)
      const { client: chatClient, options: clientOptions } = await ChatClientFactory.createChatClient();
      const systemPrompt = this.groundSystemPrompt();
      const chat = new FunctionCallingChat(chatClient, systemPrompt, this.maxOutputTokens);

      // Load chat history if specified
      if (this.inputChatHistory) {
        await chat.loadChatHistory(
          this.inputChatHistory,
          this.maxPromptTokenTarget,
          this.maxToolTokenTarget,
          this.maxChatTokenTarget
        );
      }

      // Add user prompt messages
      if (this.userPromptAdds.length > 0) {
        chat.addUserMessages(
          this.userPromptAdds,
          this.maxPromptTokenTarget,
          this.maxChatTokenTarget
        );
      }

      // Main chat loop - matching C# while loop structure
      while (true) {
        this.displayUserPrompt();
        
        const userPrompt = interactive 
          ? await this.interactivelyReadLineOrSimulateInput(this.inputInstructions, 'exit')
          : this.readLineOrSimulateInput(this.inputInstructions, 'exit');
          
        if (!userPrompt || userPrompt.trim() === '' || userPrompt === 'exit') {
          break;
        }

        // Handle chat commands (slash commands)
        const { skipAssistant, replaceUserPrompt } = await this.tryHandleChatCommand(chat, userPrompt);
        if (skipAssistant) {
          continue;
        }

        const shouldReplaceUserPrompt = !!replaceUserPrompt;
        if (shouldReplaceUserPrompt) {
          this.displayPromptReplacement(userPrompt, replaceUserPrompt!);
        }

        const giveAssistant = shouldReplaceUserPrompt ? replaceUserPrompt! : userPrompt;

        this.displayAssistantLabel();
        
        // Complete chat with streaming
        const response = await this.completeChatStreamingAsync(
          chat,
          giveAssistant,
          [],  // imageFiles - empty for now
          (messages) => this.handleUpdateMessages(messages),
          (update) => this.handleStreamingChatCompletionUpdate(update)
        );
        
        ConsoleHelpers.writeLine('\n', true);
      }

      return 0;
    } catch (error) {
      console.error('Chat command error:', error);
      return 1;
    }
  }

  private processOptions(options: any): void {
    // Process input instructions
    if (options.input) {
      this.inputInstructions.push(options.input);
    }
    if (options.inputs) {
      this.inputInstructions.push(...options.inputs);
    }
    if (options.question) {
      this.inputInstructions.push(options.question);
    }

    // Process system prompt options
    if (options.systemPrompt) {
      this.systemPrompt = options.systemPrompt;
    }
    if (options.addSystemPrompt) {
      this.systemPromptAdds.push(options.addSystemPrompt);
    }

    // Process chat history options
    if (options.inputChatHistory) {
      this.inputChatHistory = options.inputChatHistory;
    }
    if (options.outputChatHistory) {
      this.outputChatHistory = options.outputChatHistory;
    }
    if (options.chatHistory) {
      this.inputChatHistory = options.chatHistory;
      this.outputChatHistory = options.chatHistory;
    }
    if (options.continue) {
      this.loadMostRecentChatHistory = true;
    }

    // Process variables
    if (options.var) {
      this.variables = { ...this.variables, ...options.var };
    }

    // Process token limits
    if (options.maxOutputTokens) {
      this.maxOutputTokens = options.maxOutputTokens;
    }
  }

  private groundSystemPrompt(): string {
    let systemPrompt = this.systemPrompt || this.getBuiltInSystemPrompt();
    
    // Add system prompt additions
    if (this.systemPromptAdds.length > 0) {
      const additions = this.systemPromptAdds.join('\n\n');
      systemPrompt = `${systemPrompt}\n\n${additions}`;
    }

    // Process templates if enabled
    if (this.useTemplates) {
      systemPrompt = this.processTemplate(systemPrompt);
    }

    return systemPrompt.trim();
  }

  private getBuiltInSystemPrompt(): string {
    return 'You are a helpful AI assistant.';
  }

  private processTemplate(template: string): string {
    if (!template) return template;
    
    // Simple variable substitution - replace {variableName} with values
    let processed = template;
    for (const [name, value] of Object.entries(this.variables)) {
      const pattern = new RegExp(`\\{${name}\\}`, 'g');
      processed = processed.replace(pattern, value);
    }
    
    // Built-in variables
    const now = new Date();
    processed = processed.replace(/\{date\}/g, now.toISOString().split('T')[0]);
    processed = processed.replace(/\{time\}/g, now.toTimeString().split(' ')[0]);
    processed = processed.replace(/\{year\}/g, now.getFullYear().toString());
    processed = processed.replace(/\{month\}/g, (now.getMonth() + 1).toString().padStart(2, '0'));
    processed = processed.replace(/\{day\}/g, now.getDate().toString().padStart(2, '0'));
    
    return processed;
  }

  private async tryHandleChatCommand(chat: FunctionCallingChat, userPrompt: string): Promise<{ skipAssistant: boolean; replaceUserPrompt?: string }> {
    if (userPrompt.startsWith('/save')) {
      const fileName = userPrompt.substring('/save'.length).trim();
      return { skipAssistant: this.handleSaveChatHistoryCommand(chat, fileName) };
    } else if (userPrompt === '/clear') {
      return { skipAssistant: this.handleClearChatHistoryCommand(chat) };
    } else if (userPrompt === '/cost') {
      return { skipAssistant: this.handleShowCostCommand() };
    } else if (userPrompt === '/help') {
      return { skipAssistant: this.handleHelpCommand() };
    }

    return { skipAssistant: false };
  }

  private handleSaveChatHistoryCommand(chat: FunctionCallingChat, fileName?: string): boolean {
    const useDefaultFileName = !fileName || fileName.trim() === '';
    const finalFileName = useDefaultFileName ? 'chat-history.jsonl' : fileName;

    ConsoleHelpers.writeWithColor(`Saving ${finalFileName} ...`, 'yellow', true);
    chat.saveChatHistoryToFile(finalFileName);
    ConsoleHelpers.writeWithColor('Saved!\n', 'yellow', true);

    return true;
  }

  private handleClearChatHistoryCommand(chat: FunctionCallingChat): boolean {
    chat.clearChatHistory();
    this.totalTokensIn = 0;
    this.totalTokensOut = 0;
    ConsoleHelpers.writeWithColor('Cleared chat history.\n', 'yellow', true);
    return true;
  }

  private handleShowCostCommand(): boolean {
    ConsoleHelpers.writeWithColor(`Tokens: ${this.totalTokensIn} in, ${this.totalTokensOut} out\n`, 'yellow', true);
    return true;
  }

  private handleHelpCommand(): boolean {
    const helpText = `
  BUILT-IN

    /save     Save chat history to file
    /clear    Clear chat history
    /cost     Show token usage statistics
    /help     Show this help message

  EXTERNAL

    /files    List files matching pattern
    /file     Get contents of a file
    /find     Find content in files

    /search   Search the web
    /get      Get content from URL

    /run      Run a command
    /image    Add image file(s) to conversation

  PROMPTS

    No custom prompts found.

`;
    ConsoleHelpers.writeLine(helpText, true);
    return true;
  }

  private readLineOrSimulateInput(inputInstructions: string[], defaultOnEndOfInput?: string): string | null {
    if (inputInstructions.length > 0) {
      const input = inputInstructions.shift();
      if (input && input.trim() !== '') {
        ConsoleHelpers.writeLine(input);
        return input;
      }
    }
    return defaultOnEndOfInput || null;
  }

  private async interactivelyReadLineOrSimulateInput(inputInstructions: string[], defaultOnEndOfInput?: string): Promise<string | null> {
    const input = this.readLineOrSimulateInput(inputInstructions);
    if (input !== null) return input;

    // For interactive mode, we'd use readline here
    // For now, return the default
    return defaultOnEndOfInput || null;
  }

  private displayUserPrompt(): void {
    ConsoleHelpers.writeWithColor('\rUser: ', 'green');
  }

  private displayPromptReplacement(userPrompt: string, replaceUserPrompt: string): void {
    ConsoleHelpers.writeWithColor(`\rUser: ${userPrompt} => ${replaceUserPrompt}`, 'gray', true);
  }

  private displayAssistantLabel(): void {
    ConsoleHelpers.writeWithColor('\nAssistant: ', 'green');
    this.assistantResponseCharsSinceLabel = 0;
  }

  private displayAssistantResponse(text: string): void {
    if (this.assistantResponseCharsSinceLabel === 0 && text.startsWith('\n')) {
      text = text.trimStart();
    }

    ConsoleHelpers.writeWithColor(text, 'white', true);

    this.assistantResponseCharsSinceLabel += text.length;
    this.assistantResponseNeedsLF = !text.endsWith('\n');
  }

  private async completeChatStreamingAsync(
    chat: FunctionCallingChat,
    userPrompt: string,
    imageFiles: string[],
    messageCallback?: (messages: ChatMessage[]) => void,
    streamingCallback?: (update: any) => void
  ): Promise<string> {
    try {
      return await chat.completeChatStreamingAsync(
        userPrompt,
        imageFiles,
        messageCallback,
        streamingCallback
      );
    } catch (error) {
      console.error('Error during chat completion:', error);
      throw error;
    }
  }

  private handleUpdateMessages(messages: ChatMessage[]): void {
    // Handle auto-save and message updates
    // For now, just track the messages
  }

  private handleStreamingChatCompletionUpdate(update: any): void {
    // Extract token usage information if available
    if (update.usage) {
      this.totalTokensIn += update.usage.input_tokens || 0;
      this.totalTokensOut += update.usage.output_tokens || 0;
    }

    // Extract and display text content
    if (update.delta?.text || update.content) {
      const text = update.delta?.text || update.content || '';
      this.displayAssistantResponse(text);
    }
  }
}